### DISCUSSION TITLE
Resolve namespace of object stored in variable (Ruby)

### DISCUSSION THREAD

**Discussion by 0x73746F70626F74686572696E676D65** [github/codeql#16114] [Q&A]
How can I resolve `var` to `Scope1::Scope2::Object` given `var = Scope1::Scope2::Object.new`? Since this has to be interprocedural, the most obvious way seems to use the global DataFlow library and construct a configuration for this but I think it's odd that it's proven this difficult to do this a simpler, shorter way. Why isn't there a simple predicate in the API or DataFlow nodes that allows just resolving the receiver object? Not the direct receiver expression like the variable or a return value (`Object.new.method` and you want to resolve the receiver of method to the Object constant access, not the return value of new). Can anyone help?

**Comment by alexrford** [comment]
I'd personally approach this by using API graphs to find module instances (`m.trackInstance()`) and type-tracking to track where these instances can flow to, e.g:

```ql
private import codeql.ruby.typetracking.TypeTracking

private DataFlow::LocalSourceNode trackModuleInstance(DataFlow::ModuleNode m, TypeTracker t) {
  t.start() and
  result.track() = m.trackInstance()
  or
  exists(TypeTracker t2 | result = trackModuleInstance(m, t2).track(t2, t))
}

private DataFlow::Node trackModuleInstance(DataFlow::ModuleNode m) {
  trackModuleInstance(m, TypeTracker::end()).flowsTo(result)
}
```

**Reply by 0x73746F70626F74686572696E676D65** [reply]
Im sorry but the recursive part of this code is seriously confusing me and the documentation for TypeTracking is just a variant of this code. Can you explain what this code is doing at a high level? I would greatly appreciate it. Thanks.
